src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTimeoutManager.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * Object that handles management of timeout tasks across all sql server connections
+ */
+final class SQLServerTimeoutManager {
+    private static ScheduledExecutorService scheduledTimeoutTasks = Executors.newScheduledThreadPool(1,
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager"));
+    private static ExecutorService timeoutTaskWorker = Executors.newSingleThreadExecutor(
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager.TimeoutTaskWorker"));
+    private static Map<UUID, List<TimeoutCommand<?>>> timeoutCommands = new HashMap<>();
+    private final static Logger logger = Logger.getLogger("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager");
+
+    static void startTimeoutCommand(TimeoutCommand<?> timeoutCommand) {

```

Comment:
A bit hard to read due to the multiple nested runnables. Also shouldn't just blanket catch any exception and attribute the cause to a timeout. Suggested change pasted below:

```
static void startTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
    if (scheduledTimeoutTasks.isShutdown()) {
        // reset id counter for timeout commands
        TimeoutCommand.uniqueId.set(0);
        scheduledTimeoutTasks = Executors.newScheduledThreadPool(1);
        timeoutTaskWorker = Executors.newSingleThreadExecutor();
    }

    Runnable timeoutTaskRunnable = () -> {
        Future<?> timeoutTask = timeoutTaskWorker.submit(() -> {
            timeoutCommand.interrupt();
        });
        try {
            // if the timeout command takes too long to interrupt, cancel it and release
            timeoutTask.get(10, TimeUnit.SECONDS);
        } catch (TimeoutException te) {
            logger.log(Level.WARNING, "Timeout task too long, aborting", te);
        } catch (InterruptedException | ExecutionException e) {
            logger.log(Level.FINE, "Unexpected Exception occured in timeout thread.", e);
        } finally {
            releaseTimeoutCommand(timeoutCommand);
        }
    };

    timeoutCommand.setTimeoutTask(scheduledTimeoutTasks.schedule(() -> {
        timeoutTaskRunnable.run();
    }, timeoutCommand.getTimeout(), TimeUnit.SECONDS));
    addTimeoutCommand(timeoutCommand);
}
```